### PR TITLE
PLNSRVCE-1407: performance workgroup motivated metrics

### PR DESCRIPTION
**Pull request by gabemontero**
1) while @adambkaplan and I initially tried to minimize cardinality of the metrics, he and I had a conversation where we agreed that we needed to add a way to opt into greater cardinality; minimally during our performance testing; perhaps on by default long term;  added a commit which allows for adding the pipeline name as a label on the scheduled duration metric
2) added a task run level scheduled duration metric, with opt into a task name label, as an alternative to `tekton_pipelines_controller_taskruns_pod_latency_*` from upstream, where both @jhutar and I have seen odd results
3) added a new metric to highlight the gap in time between the scheduling of sequential  taskruns within a pipelinerun

### CODE REVIEW

File: collector/collector_test.go
Comment (ramessesii2) [inline collector/collector_test.go:149]: How about using `ctx` defined in line [147](https://github.com/openshift-pipelines/pipeline-service-exporter/pull/30/files#diff-66bee09f1f39294b9c301a94f017eacd9bf91ad3182c16ca6e2803252e919df7R147) to be consistent?

File: collector/collector_test.go
Comment (ramessesii2) [inline collector/collector_test.go:164]: we're missing the tests for `PipelineRunTaskRunGapCollector`.

File: collector/taskrun.go
Comment (ramessesii2) [inline collector/taskrun.go:62]: Do you think adding something like (with maybe a higher verbosity level)-
```
log.V(7).Info("Starting reconciliation", "namespace", request.Namespace, "name", request.Name)
```
will make sense or will it still be a lot of noise?

**Review by ramessesii2 — CHANGES_REQUESTED** [review]
Overall it looks good to me, a couple of comments though( I haven't tested the Exporter locally yet). 
Since we're adding new metrics, we might want to update docs as well.

File: collector/taskrun.go
Comment (gabemontero) [reply collector/taskrun.go:62]: yeah I don't think a log like this for a basic "did we get events" notification really helps

by the way, there is already a controller runtime stat that provides this:  see `controller_runtime_reconcile_total` over at https://book.kubebuilder.io/reference/metrics-reference.html

File: collector/collector_test.go
Comment (gabemontero) [reply collector/collector_test.go:149]: part of next push

File: collector/collector_test.go
Comment (gabemontero) [reply collector/collector_test.go:164]: will do, but rather than around this line of the TaskRun tests it should be added in the pipeline run tests

part of next push

**Comment by gabemontero** [comment]
OK @ramessesii2 I made updates based on your comments and packaged them in a separate commit https://github.com/openshift-pipelines/pipeline-service-exporter/pull/30/commits/53aa96e90ea91554dd4943451cb5eadf99db4622 so you can just look at the deltas motivated by your comments

Once we are agreed on the changes, I'll squash the commit, and unless anyone else wants to review, we'll merge.

**Comment by gabemontero** [comment]
oh I forgot - I'm also debating making the new gap metric optional, given its potential cardinality wrt labels

I'll handle that separately once we sort through your existing set of comments here

File: collector/collector_test.go
Comment (ramessesii2) [reply collector/collector_test.go:260]: sure, Thanks

File: collector/taskrun.go
Comment (ramessesii2) [reply collector/taskrun.go:62]: Sure, that makes sense.

**Comment by ramessesii2** [comment]
@gabemontero the changes look good to me. 
I can approve this PR if we're okay with having an updated doc later in separate PR. I think we'll need to update both `metrics-specification.md` and `README.md` if you push a change a/c to  your comment [here](https://github.com/openshift-pipelines/pipeline-service-exporter/pull/30#issuecomment-1670230700).

**Comment by gabemontero** [comment]
I'll do the doc updates with this PR @ramessesii2 thanks for the reminder

File: collector/pipelinerun.go
Comment (enarha) [inline collector/pipelinerun.go:149]: Shouldn't we return err here? I see you do the same for the taskRun.

File: collector/collector.go
Comment (gabemontero) [reply collector/collector.go:129]: good point ... removing the error rc

**Review by ramessesii2 — APPROVED** [review]
Thanks for updating docs Gabe 😺 . Everything LGTM. 
There are few places where I can see `Stonesoup` being referenced. But it's okay, I can do a separate PR to fix them.

**Review by ramessesii2 — COMMENTED** [review]
/lgtm 
Just realised that you still need to squash the commits and of course you might add things to the PR. Sorry about early approval.

**Comment by gabemontero** [comment]
> /lgtm Just realised that you still need to squash the commits and of course you might add things to the PR. Sorry about early approval.

no worries @ramessesii2 

that said, I've squashed the existing commits, assuming given the time period there was sufficient time to review them.

But yes, I next plan on making the new gap metric optional.  Given its cardinality, I'm considering not having it present by default, but then enable it first in some performance wg CI runs, and then perhaps having it available for the next KPI test in staging plus letting it soak in staging for a bit before perhaps turning it on by default.

I'll update here @ramessesii2 @enarha when the configuration bit is ready.

**Comment by gabemontero** [comment]
OK @ramessesii2 @enarha I've added the config switch as a separate commit.  PTAL when you get the chance.  Thanks.

File: docs/metrics-specification.md
Comment (ramessesii2) [inline docs/metrics-specification.md:26]: Nitpick:
```suggestion
The time taken in milliseconds between the creation of the first TaskRun and the creation of its PipelineRun, followed by the duration in milliseconds between the completion of a preceding TaskRun and the creation of the following TaskRun.  This metric current assumes sequential TaskRuns, and the parallel TaskRuns within a PipelineRun are not employed.  At this time, this metric is disabled by default, but can be enabled by setting the environment variable `ENABLE_GAP_METRIC` to `true`.
```

**Review by ramessesii2 — COMMENTED** [review]
/lgtm

File: docs/metrics-specification.md
Comment (jhutar) [reply docs/metrics-specification.md:26]: Maybe also `This metric current assumes` -> `This metric currently assumes`.

**Review by Roming22 — APPROVED** [review]
Everyone seems to be OK with the content. so I'm adding the required approval.
@gabemontero I'll let you press the merge `button`, in case I missed something still needed to be done.

**Comment by gabemontero** [comment]
I'll address the comment for metric-specification.md @Roming22 and squash the commits, then I'll merge

**Review by Roming22 — APPROVED** [review]

**Pull request merged**
